Apply all avatar classes to the link wrapper if applicable

laserlemon opened the pull request:
If an `href` argument is specified, the `AvatarComponent` wraps the `img` element in an `a` element, applying the `avatar` class to the `a` instead. These changes make sure that the other optional avatar-related classes like `avatar--small` and `circle` are also applied to the `a` (if present) rather than the `img`. Otherwise, you get a square `a` around a circle `img`.

| Before | After |
|:---:|:---:|
| <img width="58" alt="Screen Shot 2021-03-11 at 9 24 04 AM" src="https://user-images.githubusercontent.com/34264/110811613-d5c12200-8254-11eb-9c32-3a9433a1802d.png"> | <img width="58" alt="Screen Shot 2021-03-11 at 9 24 43 AM" src="https://user-images.githubusercontent.com/34264/110811617-d659b880-8254-11eb-9f04-5315e4eb5fa5.png"> |

laserlemon commented on test/components/avatar_component_test.rb:
```diff
@@ -46,7 +46,7 @@ def test_squared_avatar
 
     assert_selector("img.avatar")
-    refute_selector(".CircleBadge")
+    refute_selector(".circle")
```
It appeared that this should have been changed when AvatarComponent moved from using the CircleBadge class to circle.

manuelpuyol commented on app/components/primer/avatar_component.rb:
```diff
@@ -34,17 +34,25 @@ def initialize(src:, alt:, size: 20, square: false, href: nil, **system_argument
-      @system_arguments[:classes] = class_names(
-        system_arguments[:classes],
-        "avatar" => !href,
+      avatar_class_names = class_names(
```
I think we can keep  `system_arguments[:classes]` here, so If an user wants to add a custom class, it adds the class to the `<a>` instead of `<img>`.
At least it looks like that'd be the correct behavior since we are adding all classes to `<a>`

laserlemon replied:
I wondered about that. I'll make the change and update the `CHANGELOG`. Thanks for the quick review! 💛

manuelpuyol replied:
this is roughly what I had in mind

```rb
avatar_class_names = class_names(
  "avatar",
  system_arguments[:classes],
  "avatar--small" => size < SMALL_THRESHOLD,
  "circle" => !square
)

if @href
  @link_arguments = { href: href, classes: avatar_class_names }
  @system_arguments[:classes] = nil # otherwhise it will apply the custom classes twice
else
  @system_arguments[:classes] = avatar_class_names
end
```

manuelpuyol replied:
if we go this route, I'd ❤️  to see a test showing the custom class behavior too! :)

manuelpuyol reviewed (commented):
Thanks for the PR! I only have one suggestion there, but otherwise this looks great 😄 
Mind adding a CHANGELOG entry?

manuelpuyol commented on app/components/primer/avatar_component.rb:
```diff
@@ -34,17 +34,25 @@ def initialize(src:, alt:, size: 20, square: false, href: nil, **system_argument
-      @system_arguments[:classes] = class_names(
-        system_arguments[:classes],
-        "avatar" => !href,
+      avatar_class_names = class_names(
+        "avatar",
         "avatar--small" => size < SMALL_THRESHOLD,
         "circle" => !square
       )
+
+      if @href
+        @link_arguments = { href: href, classes: avatar_class_names }
+      else
+        @system_arguments[:classes] = class_names(
+          system_arguments[:classes],
+          avatar_class_names
+        )
+      end
```
```suggestion
      avatar_class_names = class_names(
        "avatar",
        system_arguments[:classes],
        "avatar--small" => size < SMALL_THRESHOLD,
        "circle" => !square
      )

      if @href
        @link_arguments = { href: href, classes: avatar_class_names }
      else
        @system_arguments[:classes] = avatar_class_names
      end
```

this is what I had in mind

laserlemon commented:
@manuelpuyol I addressed your feedback and tweaked the implementation to be non-destructive to the given arguments. I also added a bit of test coverage around setting link wrapper's `href`.

laserlemon commented on app/components/primer/avatar_component.rb:
```diff
@@ -33,22 +33,21 @@ def initialize(src:, alt:, size: 20, square: false, href: nil, **system_argument
         end
       else
-        render(Primer::BaseComponent.new(**@system_arguments)) { content }
+        render(Primer::BaseComponent.new(classes: @classes, **@system_arguments)) { content }
```
Shoot, I missed this in my refactor. Must add a test the exposes this!

laserlemon replied:
Fixed!

manuelpuyol commented on CHANGELOG.md:
```diff
@@ -34,6 +34,10 @@
 
     *Manuel Puyol*
 
+* Fix `AvatarComponent` to apply classes to the link wrapper if present.
```
mind moving this up? we tend to group breaking changes in the bottom

laserlemon replied:
👍

manuelpuyol reviewed (approved):
✨
